phpstan.neon.dist

```diff
@@ -0,0 +1,17 @@
+parameters:
+    level: 3
+    paths:
+        - lib
+    ignoreErrors:
+        # Making classes final as suggested would be a BC-break
+        -
+            message: '~Unsafe usage of new static\(\)\.~'
```

Comment:
This is a valid issue. Do you know if PHPStan supports anything like [`@psalm-consistent-constructor`](https://psalm.dev/docs/running_psalm/issues/UnsafeInstantiation/)?

Reply:
Good question, I asked it at https://symfony-devs.slack.com/archives/C8SFXTD2M/p1606661194462100

Reply:
I have tried using `@psalm-consistent-constructor` for Psalm, it only worked for `ArrayCollection`, not for `Criteria`…

Reply:
Works for me in both cases:
```console
> vendor/bin/psalm --show-info=true --no-cache                                                                                                                                 
Scanning files...
Analyzing files...

░░░░II

INFO: UnsafeInstantiation - lib/Doctrine/Common/Collections/ArrayCollection.php:100:16 - Cannot safely instantiate class Doctrine\Common\Collections\ArrayCollection with "new static" as its constructor might change in child classes (see https://psalm.dev/229)
        return new static($elements);


INFO: UnsafeInstantiation - lib/Doctrine/Common/Collections/Criteria.php:42:16 - Cannot safely instantiate class Doctrine\Common\Collections\Criteria with "new static" as its constructor might change in child classes (see https://psalm.dev/229)
        return new static();


------------------------------
No errors found!
------------------------------
2 other issues found.
------------------------------

Checks took 1.45 seconds and used 45.516MB of memory
Psalm was able to infer types for 99.2840% of the codebase

> git stash pop                                                                                                                                                                
On branch maintenance-work
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
   modified:   lib/Doctrine/Common/Collections/ArrayCollection.php
   modified:   lib/Doctrine/Common/Collections/Criteria.php

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (33260b3f4a197fcda746946e543f945b727f3659)

> git diff                                                                                                                                                                     
diff --git a/lib/Doctrine/Common/Collections/ArrayCollection.php b/lib/Doctrine/Common/Collections/ArrayCollection.php
index 963e3d1..6a16d33 100644
--- a/lib/Doctrine/Common/Collections/ArrayCollection.php
+++ b/lib/Doctrine/Common/Collections/ArrayCollection.php
@@ -39,6 +39,7 @@
  * @psalm-template T
  * @template-implements Collection<TKey,T>
  * @template-implements Selectable<TKey,T>
+ * @psalm-consistent-constructor
  */
 class ArrayCollection implements Collection, Selectable
 {
diff --git a/lib/Doctrine/Common/Collections/Criteria.php b/lib/Doctrine/Common/Collections/Criteria.php
index 8a5324a..d0aa960 100644
--- a/lib/Doctrine/Common/Collections/Criteria.php
+++ b/lib/Doctrine/Common/Collections/Criteria.php
@@ -10,6 +10,7 @@
 
 /**
  * Criteria for filtering Selectable collections.
+ * @psalm-consistent-constructor
  */
 class Criteria
 {

> vendor/bin/psalm --show-info=true --no-cache                                                                                                                                 
Scanning files...
Analyzing files...

░░░░░░
------------------------------
No errors found!
------------------------------

Checks took 1.45 seconds and used 45.516MB of memory
Psalm was able to infer types for 99.2840% of the codebase
```

Reply:
Ah, the `--no-cache` was key. Thanks!

Reply:
Another way to fix that without resorting to annotations would be to use an interface: https://phpstan.org/r/2857e582-2ede-41e2-82b6-ad4d6cda1227

I probably shouldn't do it in a patch release but it might be a good thing to do on the 1.7.x